notebooks/visit_data.py

```diff
@@ -0,0 +1,255 @@
+pagePath_count = {}
+for p_lag1, p in zip(train_data['pagePath_abbreviated_lag1'], train_data['pagePath_abbreviated']):
+    pagePath_predictor[(p_lag1, p)] = pagePath_predictor.get((p_lag1, p), 0) + 1
+    pagePath_count[p_lag1] = pagePath_count.get(p_lag1, 0) + 1
+
+pagePath_predictor_unflatten = unflatten(pagePath_predictor)
+
+
+def top_next_URLs(current_URL, population_fraction=0.01):
+    '''
+    purely for some data exploration - not really necessary
+    for the given URL, returns the top next URLs
+    requires the calculation of the dictionaries
+    - pagePath_predictor_unflatten
+    - pagePath_count
+    population_fraction at 0.01 works well
+    '''
+    for k, v in pagePath_predictor_unflatten[current_URL].items():
+        if v > population_fraction * pagePath_count[current_URL]:
+            print(k, v)
+
+
+# companies house
+top_next_URLs('companies-house')
+top_next_URLs('coronavirus')
+top_next_URLs('hm-revenue-customs')
+top_next_URLs('driving')
+top_next_URLs('new-national-restrictions-from-5-november')
+top_next_URLs('transition')
+top_next_URLs('vehicle-tax')
+top_next_URLs('foreign-travel-advice')
+top_next_URLs('universal-credit')
+
+# this is the probability for getting the page prediction correct by random sampling
+predictions = []
+for p_lag1, p in zip(test_data['pagePath_abbreviated_lag1'], test_data['pagePath_abbreviated']):
+    if pagePath_predictor.get((p_lag1, p), 0) == 0:
+        predictions.append(0)
+    else:
+        predictions.append(pagePath_predictor.get((p_lag1, p), 0) / pagePath_count.get(p_lag1, 0))
+
+len(predictions)
+np.mean(predictions)
+
+# to calculate the rank, we SORT and ENUMERATE the next_URL:count for each of the current_URL
+# note that in the code beneath, we are using a nested structure by calling 'unflatten()'
+pagePath_predictor_unflatten = unflatten(pagePath_predictor)
+
+pagePath_predictor_unflatten_ranked = {}
+for k, v in pagePath_predictor_unflatten.items():
+    temp_dict = {key: rank for rank, key in enumerate(sorted(v, key=v.get, reverse=True), 1)}
+    pagePath_predictor_unflatten_ranked[k] = temp_dict
+
+pagePath_predictor_ranked = flatten(pagePath_predictor_unflatten_ranked)
+
+# now we just look up the rank using the test data key (p_lag1, p) in the flattened rank dictionary
+reciprocal_rank_predictions = []
+for p_lag1, p in zip(test_data['pagePath_abbreviated_lag1'], test_data['pagePath_abbreviated']):
+    reciprocal_rank_predictions.append(1 / pagePath_predictor_ranked.get((p_lag1, p), 10000))
+
+# this is the mean reciprocal rank of correct predictions
+np.mean(reciprocal_rank_predictions)
+test_data['reciprocal_rank_predictions'] = reciprocal_rank_predictions
+# add some binary flags so we can say if the correct rank was in the top [x] predictions
+test_data['top_10'] = [1 if rrp >= 0.1 else 0 for rrp in test_data['reciprocal_rank_predictions']]
+test_data['top_5'] = [1 if rrp >= 0.2 else 0 for rrp in test_data['reciprocal_rank_predictions']]
```

Comment:
Not sure but should the thresholds be `rrp >= 0.9` and `rrp >= 0.95` respectively?